dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,211 @@
+ */
+template<typename T, bool new_axis = true, int dims = -1>
+struct TensorJoinCpu {
+  ///@{
+  /**
+   * @param in_shapes Shapes of input tensors.
+   * @param axis Axis, along which tensors will be joined. Accepts positive and negative values
+   *             (e.g. axis: -1 <=> ndims-1)
+   */
+  KernelRequirements Setup(KernelContext &ctx, span<const TensorShape<dims>> in_shapes, int axis) {
+    n_input_tensors_ = in_shapes.size();
+    orig_shapes_(in_shapes);
+    auto ndims = in_shapes[0].sample_dim();
+    DALI_ENFORCE(axis < ndims && axis > -ndims, "Incorrect axis. Actual: ", axis, ". Expected in [",
+                 -ndims + 1, ", ", ndims - 1, "] interval");
+    axis_ = axis >= 0 ? axis : ndims + axis;
+
+    {
+      const auto &ref = in_shapes[0];
+      for (int i = 1; i < n_input_tensors_; i++) {
+        DALI_ENFORCE(in_shapes[i].sample_dim() == ref.sample_dim(),
+                     "Every input shape must have the same number of dimensions.");
+        for (int j = 0; j < ref.size(); j++) {
+          if (!new_axis) {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j] || j == axis_, make_string(
+                    "Number of samples in every dimension (except the concatenated one) "
+                    "must be the same (CONCAT mode). 0-th shape at index ", j, " has dimension ",
+                    ref.shape[j], ", while ", i, "-th shape at index ", j,
+                    " has dimension ", in_shapes[i][j]));
+          } else {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j], make_string(
+                    "Number of samples in every dimension must be the same (STACK mode). "
+                    "0-th shape at index ", j, " has dimension ", ref.shape[j], ", while ", i,
+                    "-th shape at index ", j, " has dimension ", in_shapes[i][j]));
+          }
+        }
+      }
+    }
+
+    KernelRequirements kr;
+    output_shape_ = DetermineShape<new_axis>(in_shapes, axis);
+    kr.output_shapes.resize(1);
+    TensorListShape<> tmp({output_shape_});  // clang's destructor bug still haunting
+    kr.output_shapes[0] = tmp;
+    return kr;
+  }
+
+
+  KernelRequirements Setup(KernelContext &ctx, span<const InTensorCPU<T, dims>> in, int axis) {
+    std::vector<TensorShape<>> in_shapes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      in_shapes[i] = in[i].shape;
+    }
+    return Setup(ctx, make_span(in_shapes), axis);
+  }
+  ///@}
+
+  static constexpr int output_dims = (dims == DynamicDimensions ? DynamicDimensions :
+                                      (new_axis ? dims + 1 : dims));
+
+
+  /**
+   * @param out output tensor. Must be properly allocated
+   * @param in input tensors. The number of these tensors, as well as their shapes,
+   *           must match with what is provided in Setup call.
+   */
+  void Run(KernelContext &ctx, const OutTensorCPU<T, output_dims> &out,
+           span<const InTensorCPU<T, dims>> in) {
+    if (in.size() != n_input_tensors_) {
```

Comment:
How likely is that? Do we need to check it?

Reply:
On the event user provides the input to `Run`, that has more tensors than was provided in `Setup`, we'll get a segfault.  This is why I wanted to add this check

Reply:
I'd say that it is assumed that this won't happen in a well-written program. Should we make it an assert instead? (just a suggestion though)

Reply:
Again, I agree that in a well-written program it wouldn't happen. But in case it's not well-written, segfault might occur and in some really strange situation.

I guess that's an open question for us, how much error-checking we would like to have between Setup and Run calls (I don't remember discussing it before)

Reply:
That's why we have asserts, to check logic errors. If we were to check every error this way we will end up with extremely verbose code.
I think that assuming same input for Setup and Run is OK, and an assert would suffice for error checking.
I think we have this assumption all around our code base